consensus/src/state_computer.rs

```diff
@@ -473,3 +481,149 @@ async fn test_commit_sync_race() {
+        self.block_tx
+            .send(ExecuteBlockCommand {
+                block,
+                parent_block_id,
+                maybe_block_gas_limit,
+                result_tx,
+            })
+            .await
+            .expect("Failed to send block to execution pipeline.");
+
+        Box::pin(async move {
+            result_rx
+                .await
+                .map_err(|err| ExecutorError::InternalError {
+                    error: format!(
+                        "Failed to receive execution result for block {}: {:?}.",
+                        block_id, err
+                    ),
+                })?
+        })
+    }
+
+    async fn execute_stage(
+        mut block_rx: mpsc::Receiver<ExecuteBlockCommand>,
+        ledger_apply_tx: mpsc::Sender<LedgerApplyCommand>,
+        executor: Arc<dyn BlockExecutorTrait>,
+    ) {
+        while let Some(ExecuteBlockCommand {
+            block,
+            parent_block_id,
+            maybe_block_gas_limit,
+            result_tx,
+        }) = block_rx.recv().await
+        {
+            let block_id = block.block_id;
+            debug!("execute_stage received block {}.", block_id);
+            let executor = executor.clone();
+            let state_checkpoint_output = monitor!(
+                "execute_block",
+                tokio::task::spawn_blocking(move || {
+                    fail_point!("consensus::compute", |_| {
+                        Err(ExecutorError::InternalError {
+                            error: "Injected error in compute".into(),
+                        })
+                    });
+                    executor.execute_and_state_checkpoint(
+                        block,
+                        parent_block_id,
+                        maybe_block_gas_limit,
+                    )
+                })
+                .await
+            )
+            .expect("Failed to spawn_blocking.");
+
+            ledger_apply_tx
+                .send(LedgerApplyCommand {
+                    block_id,
+                    parent_block_id,
+                    state_checkpoint_output,
+                    result_tx,
+                })
+                .await
+                .expect("Failed to send block to ledger_apply stage.");
+        }
+        debug!("execute_stage quitting.");
+    }
+
+    async fn ledger_apply_stage(
+        mut block_rx: mpsc::Receiver<LedgerApplyCommand>,
+        executor: Arc<dyn BlockExecutorTrait>,
+    ) {
+        while let Some(LedgerApplyCommand {
+            block_id,
+            parent_block_id,
+            state_checkpoint_output,
+            result_tx,
+        }) = block_rx.recv().await
+        {
+            debug!("ledger_apply stage received block {}.", block_id);
+            let res = async {
+                let executor = executor.clone();
+                monitor!(
+                    "ledger_apply",
+                    tokio::task::spawn_blocking(move || {
```

Comment:
Why do we need to use spawn_blocking here and above?

Reply:
That's the way it's supposed to execute run CPU and IO intensive tasks in async environment, to avoid blocking the async event loop. (Search for `spawn_blocking` to see how we do this today for all executor calls (hopefully))